File: test/harness/java/src/test/java/analytics/Support.java
package analytics;

import com.hightouch.analytics.Analytics;
import com.hightouch.analytics.MessageInterceptor;
import com.hightouch.analytics.messages.Message;
import java.util.Map;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

final class Support {
  interface Run {
    void apply(Analytics analytics, HtEvents events);
  }

  static <T extends Message> T sendAndRead(Class<T> type, Run run) {
    BlockingQueue<Message> queue = new ArrayBlockingQueue<>(1);
    Analytics analytics =
        Analytics.builder("wk")
            .flushInterval(365, TimeUnit.DAYS)
            .flushQueueSize(Integer.MAX_VALUE)
            .messageInterceptor(
                new MessageInterceptor() {
                  @Override
                  public Message intercept(Message message) {
                    queue.offer(message);
                    return null;
                  }
                })
            .build();
    try {
      run.apply(analytics, new HtEvents(analytics));
      Message message;
      try {
        message = queue.poll(2, TimeUnit.SECONDS);
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        throw new AssertionError("interrupted waiting for " + type.getSimpleName(), e);
      }
      if (message == null) {
        throw new AssertionError("timed out waiting for " + type.getSimpleName());
      }
      return type.cast(message);
    } finally {
      analytics.shutdown();
    }
  }

  @SuppressWarnings("unchecked")
  static Map<String, Object> asMap(Object value) {
    if (!(value instanceof Map)) {
      throw new AssertionError("expected map, got " + (value == null ? "null" : value.getClass()));
    }
    return (Map<String, Object>) value;
  }

  static double asDouble(Object value) {
    if (!(value instanceof Number)) {
      throw new AssertionError("expected number, got " + value);
    }
    return ((Number) value).doubleValue();
  }

  private Support() {}
}

File: test/harness/kotlin/src/test/kotlin/analytics/Support.kt
package analytics

import com.hightouch.analytics.kotlin.core.Analytics
import com.hightouch.analytics.kotlin.core.BaseEvent
import com.hightouch.analytics.kotlin.core.IdentifyEvent
import com.hightouch.analytics.kotlin.core.Settings
import com.hightouch.analytics.kotlin.core.TrackEvent
import com.hightouch.analytics.kotlin.core.platform.Plugin
import com.hightouch.analytics.kotlin.core.utilities.toContent
import kotlinx.serialization.json.JsonObject
import java.util.UUID
import java.util.concurrent.CountDownLatch
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import kotlin.test.fail

internal object Support {
    /**
     * Events sit in StartupQueue until settings fetch finishes. Use HTTP to a
     * closed localhost port (the SDK only skips TLS for `localhost`) and wait
     * for a warmup track so tests do not race the queue.
     */
    private fun awaitReady(analytics: Analytics) {
        val ready = CountDownLatch(1)
        val plugin =
            object : Plugin {
                override val type = Plugin.Type.After
                override lateinit var analytics: Analytics

                override fun execute(event: BaseEvent): BaseEvent {
                    if (event is TrackEvent && event.event == "__ht_harness_ready") {
                        ready.countDown()
                    }
                    return event
                }
            }
        analytics.add(plugin)
        try {
            val deadline = System.nanoTime() + TimeUnit.SECONDS.toNanos(10)
            while (ready.count > 0L) {
                val remaining = deadline - System.nanoTime()
                if (remaining <= 0L) {
                    fail("analytics never left StartupQueue")
                }
                analytics.track("__ht_harness_ready")
                ready.await(
                    remaining.coerceAtMost(TimeUnit.MILLISECONDS.toNanos(100)),
                    TimeUnit.NANOSECONDS,
                )
            }
        } finally {
            analytics.remove(plugin)
        }
    }

    inline fun <reified T : BaseEvent> sendAndRead(
        crossinline block: (Analytics, HtEvents) -> Unit,
    ): T {
        val queue = LinkedBlockingQueue<BaseEvent>()
        val tag = UUID.randomUUID().toString()
        val analytics =
            Analytics("wk-$tag") {
                application = "n/a"
                collectDeviceId = false
                autoAddSegmentDestination = false
                trackApplicationLifecycleEvents = false
                flushAt = Int.MAX_VALUE
                flushInterval = Int.MAX_VALUE
                apiHost = "localhost:1"
                cdnHost = "localhost:1"
                defaultSettings = Settings()
            }
        awaitReady(analytics)
        analytics.add(
            object : Plugin {
                override val type = Plugin.Type.After
                override lateinit var analytics: Analytics

                override fun execute(event: BaseEvent): BaseEvent {
                    queue.offer(event)
                    return event
                }
            },
        )
        block(analytics, HtEvents(analytics))
        val deadline = System.nanoTime() + TimeUnit.SECONDS.toNanos(10)
        while (true) {
            val remaining = deadline - System.nanoTime()
            if (remaining <= 0L) {
                fail("expected ${T::class.simpleName} within 10s")
            }
            val payload = queue.poll(remaining, TimeUnit.NANOSECONDS) ?: continue
            if (payload is T) {
                return payload
            }
        }
    }

    fun identifyAndWait(analytics: Analytics, userId: String) {
        val done = CountDownLatch(1)
        val plugin =
            object : Plugin {
                override val type = Plugin.Type.After
                override lateinit var analytics: Analytics

                override fun execute(event: BaseEvent): BaseEvent {
                    if (event is IdentifyEvent && event.userId == userId) {
                        done.countDown()
                    }
                    return event
                }
            }
        analytics.add(plugin)
        try {
            analytics.identify(userId)
            if (!done.await(10, TimeUnit.SECONDS)) {
                fail("identify($userId) did not land")
            }
        } finally {
            analytics.remove(plugin)
        }
    }

    fun asMap(value: Any?): Map<String, Any?> {
        if (value is JsonObject) {
            return value.toContent()
        }
        if (value is Map<*, *>) {
            @Suppress("UNCHECKED_CAST")
            return value as Map<String, Any?>
        }
        throw AssertionError(
            "expected map, got ${value?.javaClass?.name ?: "null"}",
        )
    }

    fun asDouble(value: Any?): Double? {
        if (value is Number) {
            return value.toDouble()
        }
        return null
    }
}
